metricflow/dataflow/builder/source_node.py

```diff
@@ -21,19 +29,68 @@ def create_from_data_sets(self, data_sets: Sequence[DataSourceDataSet]) -> Seque
+                    data_set.data_source_reference
+                )
+            )
+            non_additive_measures = self._semantic_model.data_source_semantics.get_non_additive_measures(
                 data_set.data_source_reference
             )
+            measure_name_to_spec = {m.spec.element_name: m.spec for m in data_set.instance_set.measure_instances}
+            instance_set_with_no_measures = data_set.instance_set.transform(RemoveMeasures())
+
             # Dimension sources may not have any measures -> no aggregation time dimensions.
-            if len(aggregation_time_dimensions) == 0:
+            if len(agg_time_dim_to_measures) == 0:
                 source_nodes.append(read_node)
             else:
-                for time_dimension_reference in aggregation_time_dimensions:
-                    source_nodes.append(
-                        MetricTimeDimensionTransformNode[DataSourceDataSet](
+                for time_dimension_reference, measures in agg_time_dim_to_measures.items():
+                    measures_for_data_source = set(measures)
+
+                    # Build all source nodes from non-additive measures
+                    for non_additive_measure in non_additive_measures:
+                        if non_additive_measure not in measures_for_data_source:
+                            continue
+                        measures_for_data_source.remove(non_additive_measure)
+
+                        filter_non_additive_measure_node = FilterElementsNode[DataSourceDataSet](
+                            parent_node=read_node,
+                            include_specs=instance_set_with_no_measures.spec_set.all_specs
+                            + (measure_name_to_spec[non_additive_measure.name],),
+                        )
+
+                        non_additive_params = non_additive_measure.non_additive_dimension
+                        time_dimension_spec = TimeDimensionSpec.from_name(non_additive_params.name)
+                        window_groupings = tuple(
+                            LinklessIdentifierSpec.from_element_name(name)
+                            for name in non_additive_params.window_groupings
+                        )
+
+                        semi_additive_join_node = SemiAdditiveJoinNode[DataSourceDataSet](
+                            parent_node=filter_non_additive_measure_node,
+                            identifier_specs=window_groupings,
+                            time_dimension_spec=time_dimension_spec,
+                            agg_by_function=non_additive_params.window_choice,
+                        )
+
+                        source_nodes.append(
+                            MetricTimeDimensionTransformNode[DataSourceDataSet](
+                                parent_node=semi_additive_join_node,
+                                aggregation_time_dimension_reference=time_dimension_reference,
+                            )
```

Comment:
Ok this seems right to me but I'll defer to @plypaul on whatever he might turn up here.

From a code structure perspective, I think it'll be easier to understand the block contained in this inner for loop if a lot of this was moved into a helper function to initialize and return a MetricTimeDimensionTransformNode.